tempesta_fw/http_sticky.c

```diff
@@ -78,67 +79,70 @@ tfw_http_sticky_send_302(TfwHttpMsg *hm)
 	return 0;
 }
 
-/*
- * Find a specific non-special header field in an HTTP message.
- *
- * This function assumes that the header field name is stored
- * in TfwStr{} after an HTTP message is parsed.
- */
-static TfwStr *
-tfw_http_field_raw(TfwHttpMsg *hm, const char *field_name, size_t len)
-{
-	int i;
-
-	for (i = TFW_HTTP_HDR_RAW; i < hm->h_tbl->size; i++) {
-		TfwStr *hdr_field = &hm->h_tbl->tbl[i];
-		if (tfw_str_eq_cstr(hdr_field, field_name, len,
-				    TFW_STR_EQ_PREFIX | TFW_STR_EQ_CASEI))
-			return hdr_field;
-	}
-
-	return NULL;
-}
-
-static int
-tfw_http_field_value(TfwHttpMsg *hm, const TfwStr *field_name, TfwStr *value)
+static bool
+search_cookie(TfwPool *pool, const TfwStr *cookie, TfwStr *val)
 {
-	char *buf, *ptr;
-	size_t len;
-	TfwStr *hdr_field;
-
-	hdr_field = tfw_http_field_raw(hm, field_name->ptr, field_name->len);
-	if (hdr_field == NULL) {
-		return 0;
-	}
-
-	if (TFW_STR_PLAIN(hdr_field)) {
-		buf = (char *)hdr_field->ptr + field_name->len;
-		len = hdr_field->len - field_name->len;
-
-		while (len > 0 && isspace(*buf)) {
-			buf++;
-			len--;
-		}
-		while (len > 0 && isspace(*(buf + len - 1)))
-			len--;
+	const TfwStr *const str = &tfw_cfg_sticky.prefix;
+	const TfwStr *chunk, *next;
+	TfwStr *s = NULL;
+	enum {
+		StateName,
+		StateSearchVal,
+		StateVal,
+		StateValChunks,
+	} state = StateName;
+
+	BUG_ON(!TFW_STR_PLAIN(str));
+
+	TFW_STR_FOR_EACH_CHUNK(chunk, cookie, {
+		switch (state) {
+
+		case StateName:
+			if ((chunk->flags & TFW_STR_NAME)
+			    && tfw_str_eq_cstr(chunk, str->ptr, str->len,
```

Comment:
Let's say that data come in one byte at a time, so a cookie name will be scattered across multiple chunks. In that case this will not work as intended.